File: lib/Classroom/add_classroom.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_esclass_2/Classroom/setting_calss.dart';
import 'package:flutter_esclass_2/Data/Data.dart';
import 'package:fluttertoast/fluttertoast.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:connectivity_plus/connectivity_plus.dart';

class AddClassroom extends StatefulWidget {
  final Examset exam;
  final String username;
  final String thfname;
  final String thlname;
  const AddClassroom({
    super.key, 
    required this.exam,
    required this.username,
    required this.thfname,
    required this.thlname,
  });

  @override
  State<AddClassroom> createState() => _AddClassroomState();
}

class _AddClassroomState extends State<AddClassroom> {
  final formKey = GlobalKey<FormState>();
  String? selectedclassroom;
  late String Name_class;
  late String SubjectsID;
  late String Section_class;
  late String Room_year;
  late int Room_No;
  late int School_year;
  late String Detail;

  

  Future<void> addClassroom() async {
    var data = {
      'classroom_name': Name_class,
      'classroom_subjectsID': SubjectsID,
      'classroom_major': Section_class,
      'classroom_year': Room_year,
      'classroom_numroom': Room_No,
      'classroom_detail': Detail,
      'usert_username': widget.username,
    };

    var response = await http.post(
      Uri.parse('https://www.edueliteroom.com/connect/add_classroom.php'),
      body: json.encode(data),
      headers: {"Content-Type": "application/json"},
    );
    print(response.body);

    if (response.statusCode == 200) {
      var responseData = json.decode(response.body);
      if (responseData['status'] == 'success') {
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text('Classroom added successfully'), backgroundColor: Colors.green),
        );
        Navigator.pop(context, true); // ส่งค่ากลับไปที่หน้าหลักว่าเพิ่มสำเร็จ
      } else {
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text('Failed to add classroom: ${data['message']}'), backgroundColor: Colors.red),
        );
      }
    } else {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('Error: ${response.body}'), backgroundColor: Colors.red),
      );
    }
  }

  void showToast(String message) {
    FToast().init(context).showToast(
      child: Container(
        padding: EdgeInsets.symmetric(horizontal: 24.0, vertical: 12.0),
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(25.0),
          color: Colors.black87,
        ),
        child: Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            Icon(Icons.info, color: Colors.white),
            SizedBox(width: 12.0),
            Text(message, style: TextStyle(color: Colors.white)),
          ],
        ),
      ),
      gravity: ToastGravity.BOTTOM,
      toastDuration: Duration(seconds: 2),
    );
  }

  @override
  Widget build(BuildContext context) {
    final screenSize = MediaQuery.of(context).size;

    return AlertDialog(
      title: Center(child: Text("สร้างห้องเรียนของคุณ")),
      content: Form(
        key: formKey,
        child: SizedBox(
          height: screenSize.height * 0.7,
          width: screenSize.width * 0.5,
          child: SingleChildScrollView(
            scrollDirection: Axis.vertical,
            child: Column(
              children: [
                Container(
                  height: screenSize.height * 0.08,
                  width: screenSize.width * 0.3,
                  margin: EdgeInsets.all(10),
                  child: TextFormField(
                    decoration: InputDecoration(label: Text("วิชา", style: TextStyle(fontSize: 20))),
                    validator: (value) {
                      if (value == null || value.isEmpty) {
                        return "กรุณากรอกชื่อวิชา";
                      }
                      return null;
                    },
                    onSaved: (value) {
                      Name_class = value ?? '';
                    },
                  ),
                ),
                Container(
                  height: screenSize.height * 0.08,
                  width: screenSize.width * 0.3,
                  margin: EdgeInsets.all(10),
                  child: TextFormField(
                    decoration: InputDecoration(label: Text("รหัสวิชา", style: TextStyle(fontSize: 20))),
                    validator: (value) {
                      if (value == null || value.isEmpty) {
                        return "กรุณากรอกรหัสวิชา";
                      }
                      return null;
                    },
                    onSaved: (value) {
                      SubjectsID = value ?? '';
                    },
                  ),
                ),
                Container(
                  height: screenSize.height * 0.08,
                  width: screenSize.width * 0.3,
                  margin: EdgeInsets.all(10),
                  child: DropdownButtonFormField<String>(
                    decoration: InputDecoration(label: Text("แผนการเรียน", style: TextStyle(fontSize: 20))),
                    items: sectionOptions.map((String section) {
                      return DropdownMenuItem<String>(
                        value: section,
                        child: Text(section),
                      );
                    }).toList(),
                    onChanged: (value) {
                      setState(() {
                        Section_class = value ?? '';
                      });
                    },
                    validator: (value) {
                      if (value == null || value.isEmpty) {
                        return 'กรุณาเลือกแผนการเรียน';
                      }
                      return null;
                    },
                    onSaved: (value) {
                      Section_class = value ?? '';
                    },
                  ),
                ),
                Container(
                  height: screenSize.height * 0.08,
                  width: screenSize.width * 0.3,
                  margin: EdgeInsets.all(10),
                  child: DropdownButtonFormField<String>(
                    decoration: const InputDecoration(
                      label: Text("กรุณาเลือกชั้นปีการศึกษา", style: TextStyle(fontSize: 20)),
                    ),
                    value: selectedclassroom,
                    items: ["ชั้นมัธยมศึกษาปีที่ 4", "ชั้นมัธยมศึกษาปีที่ 5", "ชั้นมัธยมศึกษาปีที่ 6"]
                        .map((classroom) {
                      return DropdownMenuItem(
                        value: classroom,
                        child: Text(classroom),
                      );
                    }).toList(),
                    onChanged: (value) {
                      setState(() {
                        selectedclassroom = value;
                        Room_year = value ?? '';
                      });
                    },
                    validator: (value) => value == null ? 'กรุณาเลือกชั้นปีการศึกษา' : null,
                  ),
                ),
                Container(
                  height: screenSize.height * 0.08,
                  width: screenSize.width * 0.3,
                  margin: EdgeInsets.all(10),
                  child: DropdownButtonFormField<int>(
                    decoration: InputDecoration(label: Text("เลือกห้อง", style: TextStyle(fontSize: 20))),
                    items: List.generate(30, (index) {
                      int roomNumber = index + 1;
                      return DropdownMenuItem<int>(
                        value: roomNumber,
                        child: Text(roomNumber.toString()),
                      );
                    }),
                    onChanged: (value) {
                      setState(() {
                        Room_No = value!;
                      });
                    },
                    validator: (value) {
                      if (value == null) {
                        return "กรุณาเลือกหมายเลขห้อง";
                      }
                      return null;
                    },
                    onSaved: (value) {
                      Room_No = value!;
                    },
                  ),
                ),
                Container(
                  height: screenSize.height * 0.08,
                  width: screenSize.width * 0.3,
                  margin: EdgeInsets.all(10),
                  child: DropdownButtonFormField<int>(
                    decoration: InputDecoration(label: Text("ปีการศึกษา", style: TextStyle(fontSize: 20))),
                    items: List.generate(6, (index) {
                      int year = 2565 + index; // สร้างปีการศึกษา 2565 ถึง 2570
                      return DropdownMenuItem<int>(
                        value: year,
                        child: Text(year.toString()),
                      );
                    }),
                    onChanged: (value) {
                      setState(() {
                        School_year = value!;
                      });
                    },
                    validator: (value) {
                      if (value == null) {
                        return "กรุณาเลือกปีการศึกษา";
                      }
                      return null;
                    },
                    onSaved: (value) {
                      School_year = value!;
                    },
                  ),
                ),
                Container(
                  height: screenSize.height * 0.08,
                  width: screenSize.width * 0.3,
                  margin: EdgeInsets.all(10),
                  child: TextFormField(
                    decoration: InputDecoration(label: Text("รายละเอียดเพิ่มเติม", style: TextStyle(fontSize: 20))),
                    onSaved: (value) {
                      Detail = value ?? '';
                    },
                  ),
                ),
                SizedBox(height: 20),
                SizedBox(
                  height: screenSize.height * 0.07,
                  width: screenSize.width * 0.2,
                  child: FilledButton(
                    style: FilledButton.styleFrom(
                      backgroundColor: Color.fromARGB(255, 45, 124, 155),
                    ),
                    onPressed: () async {
                      var connectivityResult = await Connectivity().checkConnectivity();
                      // ignore: unrelated_type_equality_checks
                      if (connectivityResult == ConnectivityResult.none) {
                        showToast("กรุณาตรวจสอบการเชื่อมต่ออินเทอร์เน็ต");
                        return;
                      }

                      if (formKey.currentState!.validate()) {
                        formKey.currentState!.save();
                        await addClassroom();
                        Navigator.pushReplacement(
                          context,
                          MaterialPageRoute(builder: (context) => SettingCalss(
                            username: widget.username,
                            classroomMajor: '',
                            classroomName: '',
                            classroomNumRoom: '',
                            classroomYear: '',
                            thfname: widget.thfname,
                            thlname: widget.thlname,
                            
                            )),
                        );
                      }
                    },
                    child: Text("สร้างห้องเรียน", style: TextStyle(fontSize: 20)),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
      actions: [
        TextButton(
                onPressed: () {
                  Navigator.of(context).pop();
                },
                child: Text('ปิด'),
              ),
      ],
    );
  }
}

File: API/add_classroom.php
<?php
header("Content-Type: application/json");
header("Access-Control-Allow-Origin: *");
header("Access-Control-Allow-Methods: POST");

include "connect.php"; // รวมไฟล์เชื่อมต่อฐานข้อมูล

// รับข้อมูล JSON ที่ส่งมาจาก Flutter
$data = json_decode(file_get_contents("php://input"), true);

// ตรวจสอบว่ามีข้อมูลครบถ้วนหรือไม่
if (!isset($data['classroom_major']) || !isset($data['classroom_subjectsID']) || !isset($data['classroom_year']) || !isset($data['classroom_numroom']) || !isset($data['classroom_name']) || !isset($data['classroom_detail']) || !isset($data['usert_username'])) {
    echo json_encode(['status' => 'error', 'message' => 'Incomplete data']);
    exit;
}

// ดึงข้อมูลจาก JSON
$classroom_name = $data['classroom_name'];
$classroom_major = $data['classroom_major'];
$classroom_year = $data['classroom_year'];
$classroom_numroom = $data['classroom_numroom'];
$classroom_schoolyear = $data['classroom_schoolyear'];
$classroom_detail = $data['classroom_detail'];
$usert_username = $data['usert_username'];
$classroom_subjectsID = $data['classroom_subjectsID'];
// ตรวจสอบห้องเรียนที่มีข้อมูลตรงกับ 4 ฟิลด์ที่ได้รับมา
$checkSql = "SELECT * FROM classroom WHERE classroom_name = :classroom_name AND classroom_subjectsID = :classroom_subjectsID AND classroom_major = :classroom_major AND classroom_year = :classroom_year AND classroom_numroom = :classroom_numroom AND classroom_schoolyear = :classroom_schoolyear";
$checkStmt = $conn->prepare($checkSql);

// Execute คำสั่ง SQL ตรวจสอบข้อมูล
$checkStmt->execute([
    ':classroom_name' => $classroom_name,
    ':classroom_major' => $classroom_major,
    ':classroom_year' => $classroom_year,
    ':classroom_numroom' => $classroom_numroom,
    ':classroom_schoolyear' => $classroom_schoolyear,
    ':classroom_subjectsID'=> $classroom_subjectsID,
]);

// ถ้าพบข้อมูลห้องเรียนที่ตรงกับ 4 ฟิลด์นี้แล้ว
if ($checkStmt->rowCount() > 0) {
    echo json_encode(['status' => 'error', 'message' => 'Classroom already exists']);
    exit; // หยุดการทำงาน
}

// ถ้าไม่พบข้อมูลซ้ำกัน ให้ทำการเพิ่มข้อมูลห้องเรียนใหม่
$insertSql = "INSERT INTO classroom (classroom_name, classroom_subjectsID, classroom_major, classroom_year, classroom_numroom, classroom_schoolyear ,classroom_detail, classroom_type, usert_username) 
              VALUES (:classroom_name, :classroom_subjectsID, :classroom_major, :classroom_year, :classroom_numroom, :classroom_schoolyear, :classroom_detail, 0, :usert_username)";

// เตรียมคำสั่ง SQL สำหรับการเพิ่มข้อมูล
$insertStmt = $conn->prepare($insertSql);

// Execute คำสั่ง SQL
$insertStmt->execute([
    ':classroom_name' => $classroom_name,
    ':classroom_major' => $classroom_major,
    ':classroom_year' => $classroom_year,
    ':classroom_numroom' => $classroom_numroom,
    ':classroom_schoolyear' => $classroom_schoolyear,
    ':classroom_detail' => $classroom_detail,
    ':usert_username' => $usert_username,
    ':classroom_subjectsID'=>$classroom_subjectsID
]);

// ตรวจสอบผลการทำงาน
if ($insertStmt->rowCount() > 0) {
    echo json_encode(['status' => 'success']);
} else {
    echo json_encode(['status' => 'error', 'message' => 'Failed to add classroom']);
}
?>
